Add new jobs to display the failing jobs by ROS distro

j-rivero opened the pull request:
The pull request should add a new job per each ROS distro. The new job is pretty simple and only run a System Groovy script. 

The Groovy script detect which jobs from each ROS distro are failing and check two different factors to determinate which jobs to report:
  * `min_number_of_failures` : the minimum number of failing runs to report a job
  *  `min_days_failing` : the minimum days old of the first failing run.

If both conditions are met, the job will display the name and URL of the failing job.

I've tested separately the Groovy script and the XML generation of the job but not the generate_all.py script, please give an extra look into it.

j-rivero commented on ros_buildfarm/templates/snippet/builder_system-groovy_check-for-failing-jobs.xml.em:
```diff
@@ -0,0 +1,82 @@
+// Control the minimun amount of time failing
+min_days_failing = 7
+// rosdistro to search
+rosdistro = @ros_distribution
```
This empy expansion of ros_distribution is not working. I'm not sure if the format should be different or it is a problem of embedding a empy variable inside the Groovy Script. If the latest, I can think on workarounds.

dirk-thomas replied:
You called the variable `ros_distro` (https://github.com/ros-infrastructure/ros_buildfarm/pull/454/files#diff-f331a1152ef526adbb9d081d5e96a1b8R52) not `ros_distribution`. So you have to use the same name here. And since it is a string you will need quotes around it.

j-rivero replied:
arg! I changed the name just before the commit. I set it to  `ros_distro` but I can not make it to expand playing with the quotes and the @. Any idea?

dirk-thomas commented:
Since the added jobs won't affect any of the existing ones please feel free to deploy them to Jenkins to actually run them and check the result.

j-rivero commented:
> Since the added jobs won't affect any of the existing ones please feel free to deploy them to Jenkins to actually run them and check the result.

I've generated the Lunar job for testing, [here is the result](http://build.ros.org/job/lunar_failing-jobs/1/console). Note: I had to manually modify the ros_distro expansion in the groovy script.

dirk-thomas commented on ros_buildfarm/templates/snippet/builder_system-groovy_check-for-failing-jobs.xml.em:
```diff
@@ -0,0 +1,82 @@
+      return true
+  
+    // Conditions checked, reporting failing job                                  
+    println "Failing job: " + job.name + " " +  hudson.model.Hudson.instance.getRootUrl() + job.getUrl()
```
To reduce the output this could just print the url (no text and no job name). This would make it a bit more readable.

dirk-thomas commented on scripts/misc/generate_failing_jobs_job.py:
```diff
@@ -0,0 +1,59 @@
+
+def main(argv=sys.argv[1:]):
+    parser = argparse.ArgumentParser(
+        description="Generate the 'dashboard' job on Jenkins")
```
`dashboard` - copy-n-paste.

dirk-thomas commented on ros_buildfarm/templates/snippet/builder_system-groovy_check-for-failing-jobs.xml.em:
```diff
@@ -0,0 +1,82 @@
+// Control the number of failures in a row to check
+min_number_of_failures = 5
+// Control the minimun amount of time failing
+min_days_failing = 7
```
@mikaelarguedas @tfoote @wjwwood  Since you are in charge of at least one ROS distro which thresholds (in line 8 and 10) make sense from your point of view?

@mikaelarguedas Can you please look at the output of the current job (http://build.ros.org/job/lunar_failing-jobs/1/console) and check if the result makes sense.

mikaelarguedas replied:
> @mikaelarguedas Can you please look at the output of the current job (http://build.ros.org/job/lunar_failing-jobs/1/console) and check if the result makes sense.

The results look sensible to me, a bunch of dev jobs failed but don't match the thresholds of this build so were not displayed (and that's totally fine as is).

> Since you are in charge of at least one ROS distro which thresholds (in line 8 and 10) make sense from your point of view?

This reflects well the current state of Lunar. Although, I'll leave it up to @tfoote and @wjwwood to give feedback on this because I think lunar doesn't yet have enough packages (and thus failing packages) to give good estimation of what the thresholds should be

tfoote replied:
I don't think we need to wait a full week. If it's been more than 2 days it shouldn't be an intermittent build farm outage.

dirk-thomas replied:
I thought the goal was to identify "long term" failing jobs in order to remove / blacklist them. Even a week seems to be a short amount of time for maintainers to fix the problem.

tfoote replied:
I'd like to change the paradigm of leaving packages failing on the farm to catching bad releases and rolling them back. Leaving the bad releases on the farm blocks other builds and releases from being tested and can get us into a bad state. A rollback of the new release will restore the ability to build the old version, and just like leaving it broken the old version will be available for users. I can't think of a time when a package was failing for 3+days and then recovered without the maintainer doing a new release. So we might as well roll them back in the mean time. It saves resources and unblocks others. 

What would be really useful would be to know the profile of builds of one package across each arch. 
| Package | amd64 | arm64 | 
| foo | :+1:  |   :-1: | 
| bar | :-1: | :-1: |

And the other valuable thing would be to have grouping by packages in a repo as that's the unit of release. So I can judge the successrate of a release. Often regressions are correlated across multiple packages in a release. And having them grouped by repo would be valuable too.


Possibly 

REPONAME | build type1 | build type 2| ...| 
package_name1 | fail_cout | fail_count | ... |
package_name2 | fail_cout | fail_count | ... |
package_name3 | fail_cout | fail_count | ... |

Thus at a very quick glance it would be easy to say repo X has a problem let me find the last release to review. 

And we can filter it to something like 3 failed builds in a row for any package in the repo. Basically enough to keep out an intermittent failures if a build fails for something like an upstream invalidation etc.

dirk-thomas commented on ros_buildfarm/templates/snippet/builder_system-groovy_check-for-failing-jobs.xml.em:
```diff
@@ -0,0 +1,82 @@
+rosdistro = "@ros_distro"
+
+boolean is_failing_from_n_days(int min_days_failing,
+                               Job job)
```
Please avoid alignment, use this instead:

```
boolean is_failing_from_n_days(
  int min_days_failing, Job job)
```

Same below.

mikaelarguedas commented:
Looking at the output of the job, I'm wondering if jobs that are not triggered periodically should be in that list (thinking dev jobs and PR jobs). It's useful information to have,  I'm just not sure if it should be displayed as another list, be processed by another job or not considered at all

dirk-thomas commented on ros_buildfarm/templates/misc/check_failing_jobs.xml.em:
```diff
@@ -0,0 +1,45 @@
+  <triggers>
+@(SNIPPET(
+    'trigger_timer',
+    spec='@daily',
```
Lets use a specific time so that the maintainer "know" that e.g. the job has run in the night. Maybe around 5am?

dirk-thomas commented on ros_buildfarm/templates/snippet/builder_system-groovy_check-for-failing-jobs.xml.em:
```diff
@@ -0,0 +1,82 @@
+  
+  // if there is a last success build check when it was  
+  if (last_success_build)
+  {
```
Please keep the open `{` trailing the previous line within a function (for `if`, `while`, etc.).

Same below.

dirk-thomas commented on ros_buildfarm/templates/snippet/builder_system-groovy_check-for-failing-jobs.xml.em:
```diff
@@ -0,0 +1,82 @@
+failed_jobs.each{ job -> 
+   
+    if (! is_failing_from_n_days(min_days_failing, job))
+      return true
```
Please use consistent indentation (2 spaces in Groovy?).

tfoote commented:
I agree with @mikaelarguedas I don't consider the PR and dev jobs to be in the same class. We could keep listing them, but maybe segregate them to a separate list.

The state of the dev and PR jobs is the responsibility of the maintainer to monitor. Versus the binary jobs the rosdistro boss needs to pay a lot closer attention to.

j-rivero commented:
> I added some inline comment but mostly small stuff.

I think I've fixed all the issues, thanks @dirk-thomas for the review!

> I'd like to change the paradigm of leaving packages failing on the farm to catching bad releases and rolling them back.

I see the point although. @tfoote do you think that we need to make the changes under the scope of this PR?

> I agree with @mikaelarguedas I don't consider the PR and dev jobs to be in the same class. We could keep listing them, but maybe segregate them to a separate list.

I also though it the first time I saw the results. I should be able to implement it under the scope of this PR if you think that it is the way to go.

dirk-thomas commented:
> I think I've fixed all the issues

The patch looks good to me. It implements what was described in #431.

From my point of view I can see the advantage of grouping the jobs based on their type (`doc`, `devel`, `pr`, `sourcedeb`, `binarydeb`) since each group will probably be handled differently (option A). Keep in mind that this is happening in a Groovy script and the logic doesn't have any access to rosdistro, the buildfarm config files, et al. So for some of the mentioned ideas the necessary information might not be available.

Instead the result could also be grouped by the package name / repo name (depending on the job type) as follows (option B):
* split the failing job names at the first double underscore (`__`)
* order the output by the second part (that keeps jobs for the same package name / repo name (depending on the job type) together, but always print the full url as is

Option A makes sense when handling `devel` jobs differently than `release` jobs. Option B makes sense when considering package/repo specific failures. The both have the downside that they are not supporting the opposite case well.

I will leave it up to @mikaelarguedas and @tfoote to clearly describe what they would like to have (given the information available). Please describe as concretely as possible than @j-rivero can likely update the PR with that. If the request becomes "too fancy" I would suggest to merge the current state and let @mikaelarguedas and/or @tfoote implement the view they would like to have.

tfoote commented:
Here's the perfect case for it to catch. 

I checked and there's 192 regressions. But after clicking through I found that most builds are aborted and the one recent release is failing and breaking all the downstream. To find the failure I clicked through the dependency tree walking upward. I found http://build.ros.org/view/Ibin_uT64/job/Ibin_uT64__gazebo_ros__ubuntu_trusty_amd64__binary/ and created: https://github.com/ros/rosdistro/pull/15372

It's hard for me to know if that's the only failing build or just the first one that I found. 

A snapshot of the status page. 
![image](https://user-images.githubusercontent.com/447804/27557302-f0bd0e2e-5a6d-11e7-8ba8-b8d40f854e2f.png)

dirk-thomas commented:
@j-rivero The existing job has been failing multiple times. The error message indicates that this patch needs to be updated to catch those conditions.

@mikaelarguedas @tfoote Please describe the behavior you would like to see as requested earlier. Otherwise we will stick to what is implemented right now.

j-rivero commented:
> @j-rivero The existing job has been failing multiple times. The error message indicates that this patch needs to be updated to catch those conditions.

Sorry Dirk, which one you mean by "the existing job"? I got lost in the conversation.

dirk-thomas commented:
> Sorry Dirk, which one you mean by "the existing job"? I got lost in the conversation.

The job which this PR adds: http://build.ros.org/job/lunar_failing-jobs/

mikaelarguedas commented:
@j-rivero example of failure of this job recently http://build.ros.org/job/lunar_failing-jobs/15/console

As I mentioned I think that being able to list separately jobs that are the responsibility of the maintainer (dev and PR jobs) and the job requiring action from the ROS boss (binary jobs).

Then as I said I don't have enough data to advise on whether option A or option B makes more sense in general. I guess that option A would be enough for my current use case. But I leave the floor to @tfoote to give feedback on the way to go. I'm happy to get this merged and iterate on it later on if we identify changes according to our use of it.

tfoote commented:
There are a lot more features that would be great as I outlined here: https://github.com/ros-infrastructure/ros_buildfarm/pull/454#discussion_r123415232 however that's quite a bit of feature scope. I'd suggest merging this now and we can look to add those features in the future.

tfoote commented:
Sorry wrong button

j-rivero commented:
> The job which this PR adds: http://build.ros.org/job/lunar_failing-jobs/

arg .. I saw the last builds ok and thought that you mean other job. I've added a check at the only point in the code pointed by the error (getNumber method) in 52c8c63. The job is still [happy](http://build.ros.org/job/lunar_failing-jobs/21/) although the conditions that triggered the errors are probably not present.

> I'd suggest merging this now and we can look to add those features in the future.

Ok! feel free to make the requirements for them and I can take care of coding them.

dirk-thomas commented on ros_buildfarm/templates/snippet/builder_system-groovy_check-for-failing-jobs.xml.em:
```diff
@@ -0,0 +1,82 @@
+}
+
+
+// check if the N previous from referece_build_num failed
```
Spelling: `referece_build_num`

Also what is `referece_build_num` trying to reference? I couldn't find `reference_build_num` anywhere.

j-rivero replied:
ouch, outdated comment. should make some sense after 0fe10f7

dirk-thomas commented:
If there are no objections I will merge this and afterwards make a new release.

mikaelarguedas commented:
> If there are no objections I will merge this and afterwards make a new release.

:+1:
